### ISSUE TITLE
No indication (in ossec.log on agent) that logcollector is analyzing localfile entries that contain wildcards

### ISSUE THREAD

**Issue by cbj4074** [wazuh/wazuh#7245]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.0.3-4014 | logcollector | Agent | All-in-one deployment w/step-by-step installation | Ubuntu 20.04 LTS |

When I use a wildcard in a `localfile` path, the `logcollector` facility on the agent does not write log entries to `/var/ossec/logs/ossec.log` (on the agent) indicating that the files to which the wildcard resolves are being analyzed.

In other words, if the effective agent configuration is

```xml
<agent_config>
    <localfile>
        <location>/var/log/clamav/*.log</location>
        <log_format>syslog</log_format>
    </localfile>
</agent_config>
```

then `ossec.log` on the agent contains the following entries upon startup (notice that none of the logs in `/var/log/clamav` are present):

```
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/ossec/logs/active-responses.log'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/auth.log'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/syslog'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/dpkg.log'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/kern.log'.
```

However, if I retrieve the configuration via the API, like this

```
curl -k -X GET "https://localhost:55000/agents/003/config/logcollector/localfile" -H  "Authorization: Bearer $TOKEN"
```

the output indicates that the two logs inside that `/var/log/clamav` directory have, in fact, been resolved and added to the configuration, e.g.

```javascript
         {
            "file":"/var/log/clamav/clamav.log",
            "logformat":"syslog",
            "ignore_binaries":"no",
            "target":[
               "agent"
            ]
         },
         {
            "file":"/var/log/clamav/freshclam.log",
            "logformat":"syslog",
            "ignore_binaries":"no"
         },
```

If I change the effective agent configuration to reference the two logs inside this directory explicitly, forgoing the wildcards, like this

```
<agent_config>
    <localfile>
        <location>/var/log/clamav/clamav.log</location>
        <log_format>syslog</log_format>
    </localfile>
    <localfile>
        <location>/var/log/clamav/freshclam.log</location>
        <log_format>syslog</log_format>
    </localfile>
</agent_config>
```

then the agent's log reflects entries for them (amongst the other default logs, shown above):

```
2021/01/21 10:50:36 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/clamav/clamav.log'.
2021/01/21 10:50:36 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/clamav/freshclam.log'.
```

When `logcollector.debug=2` is set on the agent, the log does indeed indicate that the pattern is expanded correctly, but there is still no indication that the resolved files are being analyzed:

```
2021/01/21 13:18:50 ossec-logcollector[230651] logcollector.c:1189 at check_pattern_expand(): DEBUG: (1957): New file that matches the '/var/log/clamav/*.log' pattern: '/var/log/clamav/clamav.log'.
2021/01/21 13:18:50 ossec-logcollector[230651] logcollector.c:1064 at set_read(): DEBUG: Socket target for '/var/log/clamav/clamav.log' -> agent
2021/01/21 13:18:50 ossec-logcollector[230651] logcollector.c:1189 at check_pattern_expand(): DEBUG: (1957): New file that matches the '/var/log/clamav/*.log' pattern: '/var/log/clamav/freshclam.log'.
```

I describe this in detail in the Community channel on Slack:

https://wazuh.slack.com/archives/C0A933R8E/p1611167622003900

Possibly related to #2574 ?

**Comment by vikman90** [comment]
@cbj4074 Thank you for your detailed description of the issue. Indeed, these logs:

```
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/ossec/logs/active-responses.log'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/auth.log'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/syslog'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/dpkg.log'.
2021/01/21 12:52:17 ossec-logcollector: INFO: (1950): Analyzing file: '/var/log/kern.log'.
```

They are produced during the configuration parsing. They will appear even when such files don't exist! Just as you described, the agent prints no logs when file locations contain wildcards.

However, just as you mentioned, the agent does print debug messages when expanding wildcard patterns. The effective configuration shows that the file is being monitored, and I guess that you should be receiving those logs in the manager.

In conclusion, this issue is related to confusing logging. Nevertheless, I think that file tracking and log reporting are working properly. Should you detect any malfunction there, please let us know.

We will fix this problem and keep you posted.
Thank you.

Best regards.

**Comment by cbj4074** [comment]
@vikman90 Thanks so much for taking a look at this; I really appreciate it.

Indeed, I did not realize that `Analyzing file ...` messages were logged upon *parsing* the configuration, and not upon verifying that the files actually exist and are readable.

Just to provide some insight into my use-case and thought process here, I was using the log, rather naively, to assess whether or not my `<localfile>` directives were effective on the agent.

In the future, I will instead use the Wazuh API to retrieve the parsed and interpreted configuration for this purpose.

Given your explanation, I can certainly appreciate that there may be no (simple) means by which to resolve wildcards (or report that a given file doesn't exist, as you noted) at that stage in the configuration parsing process.

Maybe a "fix" for this confusion is more a matter of user-education, documentation, and more precise phrasing in the logging implementation.

Just a few possible ideas:

1. Instead of `Analyzing file ...`, in the log, use `Considering file ...`. I realize that this is a very subtle distinction in terminology, and in any other context, the two words mean essentially the same thing. But given the context at hand, "Analyzing" might imply that work is already being done on the file (i.e., new lines are being read and parsed), whereas "Considering" implies that work *might* be done if certain conditions are met.
2. If the shell globbing has not yet occurred at this stage, why not just print the wildcards instead, even though they may not resolve to any file(s)? So, given our real-world example here, the log would contain: `Considering file: '/var/log/clamav/*'.`
3. Document this behavior, which I am happy to do via PR, if that's helpful. I would also note some of the observations detailed in the Slack Community discussion, namely, the fact that a `<localfile>` will not be watched (and therefore not trigger alerts) if it does not exist when the `logcollector` is started.

In any event, let me know if I can be of any help on the documentation side, and I will gladly take a swing at it.

Thanks again!

**Comment by vikman90** [comment]
@cbj4074 Thank you for your response!

In my previous comment, I meant that the "Analyzing file" log was the consequence of the configuration parsing, though Logcollector print is after reading the configuration ([logcollector.c:316](https://github.com/wazuh/wazuh/blob/5bf5d50ae04d6983e38053b4efbf84851e1b95e9/src/logcollector/logcollector.c#L316)). This is a configuration sample of an unexisting file:

```xml
<localfile>
  <log_format>apache</log_format>
  <location>/var/log/lost-file.log</location>
</localfile>
```

And this is the log produced by Logcollector:
```
2021/01/27 12:21:10 wazuh-logcollector: ERROR: (1103): Could not open file '/var/log/lost-file.log' due to [(2)-(No such file or directory)].
2021/01/27 12:21:10 wazuh-logcollector: INFO: (1950): Analyzing file: '/var/log/lost-file.log'.
```

The reason to print the logs in that order is that Logcollector first needs to resolve the time pattern (like `/var/log/file-%y.log`). I agree that "Considering file" is much more appropriate for this since the behavior above is very confusing.

I think this is the best behavior for file patterns, as you suggested: 
1. "Considering file *.log"
2. "Reading/Analyzing file X" for each file matching the pattern.

For instance:
```
wazuh-logcollector: INFO: (1950): Considering file: '/var/log/*.log'.
wazuh-logcollector: INFO: (1957): Reading file: '/var/log/apport.log'.
wazuh-logcollector: INFO: (1957): Reading file: '/var/log/auth.log'.
wazuh-logcollector: INFO: (1957): Reading file: '/var/log/cloud-init-output.log'.
wazuh-logcollector: INFO: (1957): Reading file: '/var/log/cloud-init.log'.
wazuh-logcollector: INFO: (1957): Reading file: '/var/log/dpkg.log'.
wazuh-logcollector: INFO: (1957): Reading file: '/var/log/fontconfig.log'.
```

Does this meet your needs?
Best regards.

**Comment by cbj4074** [comment]
@vikman90 I see; thanks for clarifying that first point!

Beautiful! I think the proposed solution is ideal, and would address all of my logging-related concerns.

The only other question/suggestion I have is regarding item 3, from [my second reply](https://github.com/wazuh/wazuh/issues/7245#issuecomment-765603806), above.

Provided I have described the behavior accurately, i.e., a file that does not exist when the agent is started will not be monitored even if the file is created subsequently, is an important caveat to note in the documentation, in my opinion.

But now that I scour the docs more carefully, I do see that [this is, in fact, addressed](https://documentation.wazuh.com/4.0/user-manual/reference/ossec-conf/localfile.html?highlight=localfile#location):

> Wildcards can be used on Linux and Windows systems, if the log file doesn’t exist at `ossec-logcollector` start time, such log will be re-scanned after `logcollector.vcheck_files` seconds.

So, I think we're all set with just the changes you noted previously!

Thanks again for looking at this so quickly and addressing it, even though it's relatively inconsequential in the grand scheme of things!

**Comment by vikman90** [comment]
@cbj4074 Apologies, I did not reply to your 3rd question.

You're absolutely right: Logcollector checks the files every `vcheck_files` seconds:
- If any file got rotated or truncated, Logcollector reloads it.
- If `force_reload` is set to `1`, files are always reloaded even when they did not get rotated.
- Logcollector expands wildcarded paths.

There is a particular point with files referenced with no wildcards: Logcollector tries to open them every `vcheck_files` seconds, up to `open_attempts` times. If Logcollector fails to open them, then ignores them:

```
2021/02/03 12:47:51 wazuh-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/test.log'.
```

Hope it helps.
As said I'm leaving this issue open until we schedule it in a sprint.

Thank you very much for your feedback.
Best regards.

**Comment by vikman90** [comment]
Dear @cbj4074,

I confirm that this issue was fixed in 4.6.0 by PR:
- #15973

Regards.
